File: LimSeoyeon/04_DFS/BOJ_2668_숫자고르기.cpp
#include <iostream>
#include <vector>
using namespace std;
int N;
bool res[101];
bool visited[101];
int graph[101];

void init() {
	for (int i = 1; i <= N; i++) {
		visited[i] = 0;
	}
}

void copyVisit() {
	for (int i = 1; i <= N; i++) {
		if (!res[i] && visited[i]) {
			res[i] = 1;
		}
	}
}

bool DFS(int node, int st) {
	visited[node] = 1;
	int nextNode = graph[node];

	if (res[nextNode]) return false;

	if (visited[nextNode]) {
		if (nextNode == st) {
			return true;
		}
		else return false;
	}
	return DFS(nextNode, st);
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cin >> N;

	for (int i = 1; i <= N; i++) {
		int a;
		cin >> a;
		graph[i] = a;
	}
	for (int i = 1; i <= N; i++) {
		if (DFS(i, i)) copyVisit();
		init();
	}
	int cnt = 0;
	for (int i = 1; i <= N; i++) {
		cnt += res[i];
	}
	cout << cnt << '\n';

	for (int i = 1; i <= N; i++) {
		if (res[i]) {
			cout << i << '\n';
		}
	}

}

File: LimSeoyeon/04_DFS/BOJ_2668_숫자고르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    static int[] graph;
    static boolean[] visited;
    static int N;
    static boolean[] res;

    static boolean dfs(int node, int st) {
        if(visited[node]){
            if(node==st) return true;
            else return false;
        }
        visited[node] = true;
        int nextNode = graph[node];
        return dfs(nextNode, st);
    }

    static void init(){
        for(int i=1; i<=N; i++){
            visited[i]=false;
        }
    }

    static void copyVisited(){
        for(int i=1; i<=N; i++) {
            if (visited[i]&&!res[i]){
                res[i]=true;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(bf.readLine());
        //배열 선언
        graph = new int[N+1];
        visited = new boolean[N+1];
        res= new boolean[N+1];
        for(int i=1; i<=N; i++){
            graph[i]= Integer.parseInt(bf.readLine());
        }
        for(int i=1; i<=N; i++){
            if(dfs(i,i)) copyVisited();
            init();
        }
        int cnt=0;
        for(int i=1; i<=N; i++){
            if(res[i]) cnt++;
        }
        System.out.println(cnt);
        for(int i=1; i<=N; i++){
            if(res[i]) {
                System.out.println(i);
            }
        }
    }

}
